jina/drivers/__init__.py

```diff
@@ -159,7 +159,76 @@ def __getstate__(self):
         return d
 
 
-class BaseExecutableDriver(BaseDriver):
+class BaseRecursiveDriver(BaseDriver):
```

Comment:
@JoanFM what do you think about this? Almost all drivers in this PR are refactored based on this.

Reply:
The intention of this class, is to define at which depth level this drivers will apply the executor `method`?

Reply:
Some of our drivers are working with `docs`. `BaseRecursiveDriver` provides helper function so that one can apply the function on all level of `docs`, without writing recursion wrapper.

Example
```python
class KVIndexDriver(BaseIndexDriver):
    """Serialize the documents/chunks in the request to key-value JSON pairs and write it using the executor
    """

    def apply_all(self, docs: Iterable['jina_pb2.Document'], *args, **kwargs):
        self.exec_fn(docs)




class SegmentDriver(BaseCraftDriver):
    """Segment document into chunks using the executor
    """

    def __init__(self, level_names: List[str] = None, *args, **kwargs):
        super().__init__(*args, **kwargs)
        if isinstance(level_names, list) and (self._depth_end - self._depth_start + 1) != len(self.level_names):
            self.level_names = level_names
        elif level_names is None:
            pass
        else:
            raise ValueError(f'bad level names: {level_names}, the length of it should match the recursive depth + 1')

        # for adding new chunks, preorder is safer
        self.recursion_order = 'pre'
        self.counter = RandomUintCounter()

    def apply(self, doc: 'jina_pb2.Document', *args, **kwargs):
        _args_dict = pb_obj2dict(doc, self.exec.required_keys)
        ret = self.exec_fn(**_args_dict)
        if ret:
            for r in ret:
                c = doc.chunks.add()
                self.set_doc_attr(c, r, {'length', 'id', 'parent_id', 'level_depth', 'mime_type'})
                c.length = len(ret)
                c.id = next(self.counter)
                c.parent_id = doc.id
                c.level_depth = doc.level_depth + 1
                c.mime_type = doc.mime_type
            doc.length = len(ret)
        else:
            self.logger.warning(f'doc {doc.id} at level {doc.level_depth} gives no chunk')

```

Reply:
what is an example case where we might use different than (0, 0)?

Reply:
In this **SegmentDriver**, if I pass `depth_start` = 0 and `depth_end`=2, it will segment any **document** in **chunks** and ithse **chunks** will be segmented again?

Reply:
> In this SegmentDriver, if I pass depth_start = 0 and depth_end=2, it will segment any document in chunks and ithse chunks will be segmented again?

Yes, it recursively segmented all the way to `depth_end`, that's why `recursion_order` must be set to `pre`

Reply:
And do you have a useace for 'post_order` usage?

Reply:
> @JoanFM what do you think about this? Almost all drivers in this PR are refactored based on this.

In general, the PR looks great, it offers great flexibility avoiding the differentiation of chunks and documents.

Reply:
postoder traverse is often used in delete and reducing. I haven't touched the `Chunk2DocRankDriver` yet, but you can imagine, `Chunk2DocRankDriver` is basically aggregating the score at `k`th level to `k-1`th level, until we reached `level_depth=0`. In this case, `post` is required

Reply:
Ok, I see, I have a much clearer view now! Thanks!

Reply:
i'm excited about the potentials that this PR can enable